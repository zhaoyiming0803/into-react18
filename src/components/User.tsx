import * as React from 'react'
import { Dispatch } from 'react'
import { connect } from 'react-redux'

import { RootState } from '@/reducer/index'
import { addUser } from '../action/index'

import { Button } from 'antd-mobile'
import { IUser } from '@/types/index'

interface IProps {
  count: number
  userList: IUser[]
  addUser: (user: IUser) => void
}

function User<T extends IProps> (props: T) {
  const addUser = () => {
    props.addUser({
      name: 'user' + Math.random()
    })
  }
  return <>
    {
      props.userList.map((user: IUser, index: number) => (
        <div key={index} className="user">{user.name}</div>
      ))
    }
    <Button onClick={() => addUser()} className='add-user'>add user</Button>
  </>
}

const mapStateToProps = (state: RootState) => ({
  userList: state.user.userList
})

interface Action {
  type: string
  payload: IUser
}

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  addUser: (user: IUser) => dispatch(addUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(User)
