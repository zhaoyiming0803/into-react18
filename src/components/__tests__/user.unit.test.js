// https://reactjs.org/docs/test-utils.html

import React from 'react'
import renderer from 'react-test-renderer'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import User from '../User'

import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'

const mockStore = configureStore([])

describe('test User component', () => {
  let store
  let component
  let container

  beforeEach(() => {
    // mock store
    store = mockStore({
      user: {
        userList: [{
          name: 'zhangsan'
        }]
      }
    })

    // setup a DOM element as a render target
    container = document.createElement('div')
    document.body.appendChild(container)
  })

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container)
    document.body.removeChild(container)
    container = null
  })

  test('render successfully', () => {
    component = renderer.create(
      <Provider store={store}>
        <User />
      </Provider>
    ).toJSON()
    expect(component).toMatchSnapshot()
  })

  test('on click to add user', () => {
    // get a hold of the button element, and trigger some clicks on it
    act(() => {
      render(
        <Provider store={store}>
          <User />
        </Provider>,
        container
      )
    })
    const user = document.querySelector('.user')
    expect(user.innerHTML).toBe('zhangsan')
  })
})
