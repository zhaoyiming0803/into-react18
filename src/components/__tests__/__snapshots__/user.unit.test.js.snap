// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User component render successfully 1`] = `
Array [
  <div
    className="user"
  >
    zhangsan
  </div>,
  <a
    aria-disabled={false}
    className="am-button add-user"
    onClick={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
  >
    <span>
      add user
    </span>
  </a>,
]
`;
